Windows updates

ericmehl opened the pull request:
This pull request updates a few things for Windows builds, most notably updating builds to the Windows equivalent of GafferHQ Dependencies 2.1.1 and disabling permissive mode to ensure MSVC doesn't allow non-standards compliant builds.

I temporarily disabled IECoreUSD by ensuring the library can't be found. I have a follow-up PR ready to submit once this one is reviewed and I can make sure it's compatible with any changes we make along the way.

### Breaking Changes ###

- None

### Checklist ###

- [X] I have read the [contribution guidelines](https://github.com/ImageEngine/cortex/blob/master/CONTRIBUTING.md).
- [X] I have updated the documentation, if applicable.
- [X] I have tested my change(s) in the test suite, and added new test cases where necessary.
- [X] My code follows the Cortex project's prevailing coding style and conventions.
- [X] If my code made breaking changes, I applied the **pr-majorVersion** label to this PR.

johnhaddon commented on .github/workflows/main.yml:
```diff
@@ -1,6 +1,7 @@
 name: CI
 
 on:
+  workflow_dispatch:
```
Neat! I didn't know about that. Could be handy for all sorts of things, especially with the addition of `inputs:`.

johnhaddon commented on SConstruct:
```diff
@@ -1208,6 +1208,7 @@ else:
 			"/permissive-", # disable permissive mode, which also enables standard compliant two phase name lookup
 			# required in when permissive mode is off, for defining constants like M_PI used by OpenVDB
 			"/D_USE_MATH_DEFINES",
+			"/Zc:externC-",  # suppress errors from Boost::interprocess Windows SDK mismatches
```
I don't think we actually need to use `interprocess` any more, so I've opened a PR to remove it : https://github.com/ImageEngine/cortex/pull/1106. I suggest we merge that and drop this commit.

johnhaddon reviewed (changes requested):
Thanks Eric! Comments inline as usual...

andrewkaufman reviewed (changes requested):
Just leaving a reminder for myself to come back and test this at IE once the feedback has been addressed

andrewkaufman commented on .github/workflows/main.yml:
```diff
@@ -76,7 +77,7 @@ jobs:
             buildType: RELEASE
             options: config/azure/options.windows
-            dependenciesURL: "https://github.com/hypothetical-inc/gafferDependencies/releases/download/0.54.1.0/gafferDependencies-0.54.1.0-windows-msvc2017.zip"
+            dependenciesURL: "https://github.com/hypothetical-inc/gafferDependencies/releases/download/2.1.0/gafferDependencies-2.1.1-Python2-windows.zip"
```
This has a 2.2.1 zip within a 2.1.0 folder. Is that a type or expected?

ericmehl replied:
Good catch, that's a typo. I updated the build tag to be 2.1.1 and updated the Github workflow in 90adede9480e8cd44b22f9431443c55946a25a1f

ericmehl commented on config/installDependencies.py:
```diff
@@ -90,6 +90,16 @@
 	os.system( "tar xf %s -C %s --strip-components=1" % ( archiveFileName, args.dependenciesDir ) )
 else :
 	subprocess.check_output( "7z x %s -o%s -aoa -y" % ( archiveFileName, args.dependenciesDir ) )
+	# 7z (and zip extractors generally) don't have an equivalent of --strip-components=1
+	# Copy the files up one directory level to compensate
+	for p in glob.glob(
+		os.path.join(
+			args.dependenciesDir.replace( "/", "\\" ),
+			os.path.splitext( args.archiveURL.split( "/" )[-1] )[0],
+			"*"
+		)
+	):
+		shutil.move( p, args.dependenciesDir )
```
Exactly right, the latest dependencies build was pretty old and wasn't using the packager in `build.py`, it was just a zip file I put together after building. That zip file happened to have `include`, `lib`, etc. in the root of the archive, not under a subdirectory.

ericmehl commented on SConstruct:
```diff
@@ -1049,6 +1049,24 @@ env.Append(
 
+# Take a dictionary of the form {"old": "new", "old2": "new2",...}
+# and perform platform-specific substitutions on the given file
+def substituteFileText( subEnv, sourceFile, destFile, subs ) :
```
That simplifies it nicely, thanks for pointing it out. I reworked Sconstruct to use that builder in eadd22be37fb2f95dd861a217e33fd9d3e1fe435.

johnhaddon replied:
Thanks - I've added one new comment inline, since the docs builds have gone awry due to a problem in the original commit.

ericmehl replied:
I added the Doxygen build command and rebased down to here: bdbc065

ericmehl commented on SConstruct:
```diff
@@ -1049,6 +1049,24 @@ env.Append(
+		sedSubstitutions = "s/{}/{}/g".format( subs.keys()[0], subs[subs.keys()[0]] )
+		for i in range( 1, len(subs.keys() ) ) :
+			sedSubstitutions += "; s/{}/{}/g".format( subs.keys()[i], subs[subs.keys()[i]] )
+		return subEnv.Command( destFile, sourceFile, "sed \"" + sedSubstitutions + "\" $SOURCE > $TARGET" )
```
This is also neatly solved with `Substfile` in eadd22be37fb2f95dd861a217e33fd9d3e1fe435.

ericmehl commented on config/azure/options.windows:
```diff
@@ -22,11 +22,10 @@ PYTHON_LINK_FLAGS = ""
 
-BOOST_INCLUDE_PATH = deps + "\\include\\boost-1_61"
-BOOST_LIB_SUFFIX = "-vc140-mt-1_61"
+BOOST_INCLUDE_PATH = deps + "\\include\\boost-1_68"
+BOOST_LIB_SUFFIX = "-vc141-mt-x64-1_68"
```
It's probably pulling in the old default dependencies. If you are going to remove it soon anyhow, I may as well pull out the Windows part. That is done in 32b2ff123684313b7f7a79a9dd5bf00667ba0642

ericmehl commented:
Great! I think I have everything addressed except for the `interprocess` part which I can rebase onto when #1106  is in master.

It looks like this build is failing on macos with the same error about not finding `scons` as #1106 as well. Any clues on that?

johnhaddon commented:
> It looks like this build is failing on macos with the same error about not finding scons as #1106 as well. Any clues on that?

I don't think it's related to your changes, as we're getting the same issues on the Gaffer repo. GitHub seem to be mucking about with their Mac workers this week. I spent a bit of time trying to fix it yesterday, but gave up when I concluded that sometimes scons was there, and sometimes it wasn't, regardless of what I did. I'm going to leave it for this week and revisit when whatever they are doing has stabilised. Definitely no need to hold up your PR over it.

ericmehl commented:
Sounds good on the macos front for now. I rebased to clean up the typo of the dependencies and squished the file substitution / doxygen fix into a single commit. When #1106 gets merged I can polish this one off.

johnhaddon commented:
Thanks Eric, LGTM. Andrew has merged #1106 now, so if you can rebase without bdbc065 then I think we're good to merge.

ericmehl commented:
Yarg, IECoreAppleseed uses `interprocess` too so I still need to address that. I think a better alternative to my previous command line switch is to get Github workflows to build with the Visual Studio 2017 equivalent instead of 2019 that it uses now. Getting that to work is turning out to be tricky, I'll have a deeper dive tomorrow.

johnhaddon commented:
> Yarg, IECoreAppleseed uses interprocess too so I still need to address that.

Can you point me to where? Just did a quick grep and didn't come up with anything.

The command line switch workaround is fine for me if it's the most expedient path at the moment. I just preferred to remove the other interprocess usage because there was clearly a better alternative in that case.

ericmehl commented:
> Can you point me to where? Just did a quick grep and didn't come up with anything.

It wasn't in IECoreAppleseed directly but in Appleseed itself [src/appleseed/foundation/platform/atomic.h](https://github.com/appleseedhq/appleseed/blob/master/src/appleseed/foundation/platform/atomic.h) where a number of functions are inlined so they get pulled into IECoreAppleseed.

> The command line switch workaround is fine for me if it's the most expedient path at the moment. 

I think that's going to be the route we have to go. I pushed a fix that I'm pretty happy with now. Short version is I put the `/Zc:externC-` flag into the `CXXFLAGS` of the options file for Windows to disable the error. This required also putting `/permissive-` in the `CXXFLAGS` which I think is fine.

If you like the long version:
1. `/Zc:externC-` is only in MSVC 2019 and I'd prefer to stick to the VFX reference platform recommendation of using 2017 if possible. It doesn't fail anything on 2017 but every build command will say "unrecognized compiler option" which I find to be a little untidy.
2. I couldn't seem to coerce Github Workflows to using MSVC 2017 reliably - it seems to sometimes pull in some 2019 headers that cause builds to fail.
3. If you include `/Zc:externC-` _before_ `/permissive-` it doesn't have any effect. I'm guessing because `/permissive-` is a wrapper of a bunch of `/Zc:` compatibility options and the `/Zc:externC-` get clobbered. This means that we can't just set `/Zc:externC-` as a `CXXFLAG` because that would put it before `/permissive-`.
4. So I reworked the commit adding `/permissive-` and put it (and related switch) in the Scons options file rather than directly in the hard-coded build options. This way we can include `/Zc:externC-` in the correct order but allow people building in MSVC 2017 to leave it out.

Hopefully now it is sorted?(!)

johnhaddon commented:
> Hopefully now it is sorted?(!)

Thanks Eric! And happy new year!

> Just leaving a reminder for myself to come back and test this at IE once the feedback has been addressed

Did you want to take one more look @andrewkaufman? (And happy new year!)

ericmehl commented:
> Thanks Eric! And happy new year!

My pleasure, happy New Year!

andrewkaufman commented:
It works fine here (though requires is updating our ancient SCons version...

andrewkaufman commented:
Do we need to add `doc/config/Doxyfile` to the `.gitignore` list?

ericmehl commented:
> Do we need to add `doc/config/Doxyfile` to the `.gitignore` list?

Right! We should ignore that. https://github.com/ImageEngine/cortex/pull/1105/commits/c5983f0067872d83767d31d72cd0c6d1c6917061 adds it to the .gitignore list.


> (though requires is updating our ancient SCons version...

Hopefully that's not causing any major hassles?

andrewkaufman commented:
> Hopefully that's not causing any major hassles?

It should be fine.

I noticed there is now a Merge into master commit as well, was that intentional?

ericmehl commented:
Yes, when I updated just now I had a merge conflict that was a result of 676e45e6fa79aea8131d7c58f4dad19d35c063a8 which came from John's PR #1111.

I fixed it in https://github.com/ImageEngine/cortex/pull/1105/commits/aa813db2ddbeea742493d7a37bfe180375724a82

Github gave me the option of editing it in the browser, which seems to have resulted in that merge from master. Would you rather I rebase on to the latest master and push instead?

andrewkaufman commented:
> Would you rather I rebase on to the latest master and push instead?

That'd be great, thanks.

ericmehl commented:
Ok, I think that should take care of it.
